### PR TITLE
ODH argo 2.7

### PR DESCRIPTION
**Pull request by nakfour**
This adds Argo v 2.7 to ODH manifest. I called it odhargo to differentiate from KF Argo. Need to also state both cannot be installed together. 
1. Go to the portal and create an example workflow, you need to add the serviceaccount to the workflow that has the correct permissions
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: hello-world-
spec:
  serviceAccountName: argo-workflow
  entrypoint: whalesay
  templates:
  - name: whalesay
     container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["hello world"]
```
2. Here is another more elaborate workflow you can try, however you do need to install a secret first similar to the one below.
```
apiVersion: v1
kind: Secret
metadata:
  name: keysecret
type: Opaque
data:
  accesskey: xxx
  secretkey: xxx
```
```
# The following workflow executes a multi-root workflow
#
#   A   B
#  / \ /
# C   D
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: frauddetection
spec:
  serviceAccountName: argo-workflow
  entrypoint: fraud-workflow
  serviceAccountName: argo-workflow
  volumes:
  - name: workdir
    emptyDir: {}
  templates:
  - name: echo
    inputs:
      parameters:
      - name: message
    container:
      env:
      - name: ACCESS_KEY_ID
        valueFrom:
          secretKeyRef:
            name: keysecret
            key: accesskey
      - name: SECRET_ACCESS_KEY
        valueFrom:
          secretKeyRef:
            name: keysecret
            key: secretkey
      - name: S3_ENDPOINT_URL
        value: "http://s3-us-west-2.amazonaws.com/nakfour"
      image: alpine:3.7
      command: [echo, "{{inputs.parameters.message}}"]
  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["sleep 1; echo -n true > /tmp/hyper.txt"]
  - name: volumes-emptydir-example
    container:
      image: debian:latest
      command: ["/bin/bash", "-c"]
      args: ["
        vol_found=`mount | grep /mnt/vol` && \
        if [[ -n $vol_found ]]; then echo \"Volume mounted and found\"; else echo \"Not found\"; fi; sleep 1; echo -n true > /mnt/vol/hyper.txt"]
      volumeMounts:
      - name: workdir
        mountPath: /mnt/vol
    outputs:
      parameters:
      - name: should-hyper
        valueFrom:
          path: /mnt/vol/hyper.txt
  - name: fraud-workflow
    dag:
      tasks:
      - name: Read-Data-AWS
        template: echo
        arguments:
          parameters: [{name: message, value: Read-Data-AWS}]
      - name: Read-Data-Ceph
        dependencies:
        template: echo
        arguments:
          parameters: [{name: message, value: Read-Data-Ceph}]
      - name: Transform-Data
        dependencies: [Read-Data-Ceph,Read-Data-AWS]
        template: volumes-emptydir-example
      - name: Hyper-Tuning
        dependencies: [Transform-Data]
        template: echo
        when: "{{tasks.Transform-Data.outputs.parameters.should-hyper}} == true"
        arguments:
          parameters: [{name: message, value: Hyper-Tuning}]
      - name: Train-Model
        dependencies: [Transform-Data,Hyper-Tuning]
        template: echo
        arguments:
          parameters: [{name: message, value: Train-Model}]
      - name: Validate-Model
        dependencies: [Train-Model]
        template: echo
        arguments:
          parameters: [{name: message, value: "{{item}}"}]
        withItems:
        - hello world
        - goodbye world
      - name: Publish-Model
        dependencies: [Validate-Model]
        template: echo
        arguments:
          parameters: [{name: message, value: Publish-Model}]
```

### CODE REVIEW

**Comment by vpavlin** [comment]
Since we talked about Marek doing this and I reached out to him and he does not have time, I also started working on this - https://github.com/vpavlin/odh-manifests/commit/40c432e53a9d9cd356ffd855939a5f138a520c46

I don't really care which work survives, but what I would like to get the customization that is available in Kubeflow - so that we get as close to KF as possible with this component - at that point we could start experimenting with KF PIpelines and the new version of Argo etc.

@nakfour Do you plan to bring this closer to KF Argo Kustomize manifests?

**Comment by nakfour** [comment]
@vpavlin yes I think it is pretty close, however Argo 2.7 brings a lot of new features and a great new Portal. Let me know what you want added that is similar to KF?

**Comment by vpavlin** [comment]
What I was mainly interested in is the configMup configuration: https://github.com/kubeflow/manifests/blob/master/argo/base/config-map.yaml

It allows you to provide secrets for S3 credentials for artifacts. I am not sure how important it is for us, but the idea was to get as close to KF Argo as possible and see if we can start verifying KF Pipelines on newer argo and maybe file some issues

But feel free to do this in a follow up PR if you want to keep this one cleaner - just let me know and I will create an issue for that

**Comment by vpavlin** [comment]
Can you also add cluster roles then? Maybe as an overlay? 

But as I said - we can do S3 config and cluster roles in separate PRs - that might help us to merge this PR quicker

**Comment by crobby** [comment]
I installed and tried the first super simple example given, but when I try to submit it, I get the following error:  `template reference exceeded max depth (10)`

I am able to run other examples though, so I suspect it's something about the example given here.

**Comment by crobby** [comment]
> This adds Argo v 2.7 to ODH manifest. I called it odhargo to differentiate from KF Argo. Need to also state both cannot be installed together.
> 
>     1. Go to the portal and create an example workflow, you need to add the serviceaccount to the workflow that has the correct permissions
> 
> 
> ```
> apiVersion: argoproj.io/v1alpha1
> kind: Workflow
> metadata:
>   generateName: hello-world-
> spec:
>   serviceAccountName: argo-workflow
>   entrypoint: whalesay
>   templates:
>   - name: whalesay
> ```
> 
>     1. Here is another more elaborate workflow you can try
> 
> 
> ```
> # The following workflow executes a multi-root workflow
> #
> #   A   B
> #  / \ /
> # C   D
> apiVersion: argoproj.io/v1alpha1
> kind: Workflow
> metadata:
>   generateName: frauddetection
> spec:
>   serviceAccountName: argo-workflow
>   entrypoint: fraud-workflow
>   serviceAccountName: argo-workflow
>   volumes:
>   - name: workdir
>     emptyDir: {}
>   templates:
>   - name: echo
>     inputs:
>       parameters:
>       - name: message
>     container:
>       env:
>       - name: ACCESS_KEY_ID
>         valueFrom:
>           secretKeyRef:
>             name: keysecret
>             key: accesskey
>       - name: SECRET_ACCESS_KEY
>         valueFrom:
>           secretKeyRef:
>             name: keysecret
>             key: secretkey
>       - name: S3_ENDPOINT_URL
>         value: "http://s3-us-west-2.amazonaws.com/nakfour"
>       image: alpine:3.7
>       command: [echo, "{{inputs.parameters.message}}"]
>   - name: whalesay
>     container:
>       image: docker/whalesay:latest
>       command: [sh, -c]
>       args: ["sleep 1; echo -n true > /tmp/hyper.txt"]
>   - name: volumes-emptydir-example
>     container:
>       image: debian:latest
>       command: ["/bin/bash", "-c"]
>       args: ["
>         vol_found=`mount | grep /mnt/vol` && \
>         if [[ -n $vol_found ]]; then echo \"Volume mounted and found\"; else echo \"Not found\"; fi; sleep 1; echo -n true > /mnt/vol/hyper.txt"]
>       volumeMounts:
>       - name: workdir
>         mountPath: /mnt/vol
>     outputs:
>       parameters:
>       - name: should-hyper
>         valueFrom:
>           path: /mnt/vol/hyper.txt
>   - name: fraud-workflow
>     dag:
>       tasks:
>       - name: Read-Data-AWS
>         template: echo
>         arguments:
>           parameters: [{name: message, value: Read-Data-AWS}]
>       - name: Read-Data-Ceph
>         dependencies:
>         template: echo
>         arguments:
>           parameters: [{name: message, value: Read-Data-Ceph}]
>       - name: Transform-Data
>         dependencies: [Read-Data-Ceph,Read-Data-AWS]
>         template: volumes-emptydir-example
>       - name: Hyper-Tuning
>         dependencies: [Transform-Data]
>         template: echo
>         when: "{{tasks.Transform-Data.outputs.parameters.should-hyper}} == true"
>         arguments:
>           parameters: [{name: message, value: Hyper-Tuning}]
>       - name: Train-Model
>         dependencies: [Transform-Data,Hyper-Tuning]
>         template: echo
>         arguments:
>           parameters: [{name: message, value: Train-Model}]
>       - name: Validate-Model
>         dependencies: [Train-Model]
>         template: echo
>         arguments:
>           parameters: [{name: message, value: "{{item}}"}]
>         withItems:
>         - hello world
>         - goodbye world
>       - name: Publish-Model
>         dependencies: [Validate-Model]
>         template: echo
>         arguments:
>           parameters: [{name: message, value: Publish-Model}]
> ```

The second example has the serviceAccountName listed twice, which causes it to be invalid.

File: odhargo/base/cronworkflow-crd.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# This is an auto-generated file. DO NOT EDIT
```
Comment (crobby) [inline odhargo/base/cronworkflow-crd.yaml:1]: Should the CRDs go into a "cluster" dir?

File: odhargo/base/workflow-crd.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (crobby) [inline odhargo/base/workflow-crd.yaml:1]: Should the CRDs go into a "cluster" dir?

File: odhargo/base/workflowtmp-crd.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (crobby) [inline odhargo/base/workflowtmp-crd.yaml:1]: Should the CRDs go into a "cluster" dir?

**Comment by nakfour** [comment]
@crobby I added the secret that needs to be installed before the second example

File: odhargo/base/workflow-crd.yaml (hunk 2 above)
Comment (nakfour) [reply odhargo/base/workflow-crd.yaml:1]: @crobby is that what we are doing? If so I can move them

Comment (crobby) [reply odhargo/base/workflow-crd.yaml:1]: Yeah, anything that would be a "cluster-admin" sort of action goes to "cluster".

**Comment by nakfour** [comment]
@crobby for the template error, you had it correct, the example I gave was incomplete, I just updated it

**Comment by crobby** [comment]
> @crobby I added the secret that needs to be installed before the second example

Not sure it helps me run the example without having the data expected to host on my own S3 or a working key/secret for the one in the workflow.

**Comment by nakfour** [comment]
Not sure I understand your point @crobby ? To use an S3 interface is up to the user and the best practices for OCP is to use a secret for the credentials. We cannot assume all users will use S3.

**Comment by crobby** [comment]
> Not sure I understand your point @crobby ? To use an S3 interface is up to the user and the best practices for OCP is to use a secret for the credentials. We cannot assume all users will use S3.

My point is that for me to be able to run that example, I need more info...like where I'd find whatever data is expected in the workflow so I can set up my data source.

**Comment by nakfour** [comment]
@crobby the workflow is an example, it does not use any data and is just there as an example to show where to place the S3 settings, all the containers are just shells, they don't execute on any data from S3

**Comment by nakfour** [comment]
@crobby @vpavlin I added the cluster folder for the crds

**Comment by nakfour** [comment]
@crobby @vpavlin did the rebase.

**Comment by crobby** [comment]
1 more rebase and I'll merge it immediately, if not sooner.

**Comment by nakfour** [comment]
@vpavlin @crobby rebased again

**Pull request merged**